cmd/gcp-controller-manager/kubelet_readonly_csr_approver.go

```diff
@@ -0,0 +1,322 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+
+	authorization "k8s.io/api/authorization/v1"
+	capi "k8s.io/api/certificates/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cloud-provider-gcp/pkg/csrmetrics"
+	"k8s.io/klog/v2"
+	certutil "k8s.io/kubernetes/pkg/apis/certificates/v1"
+	"k8s.io/kubernetes/pkg/controller/certificates"
+)
+
+const (
+	kubeletReadonlyCSRSignerName         = "gke.io/gke-kubelet-readonly-client"
+	kubeletReadonlyCSRRequestMetrics     = "kubelet_readonly_csr_request"
+	podNameKey                           = "authentication.kubernetes.io/pod-name"
+	kubeletAPILimitedReaderAnnotationKey = "autopilot.gke.io/kubelet-api-limited-reader"
+)
+
+var (
+	kubeletReadonlyUsageAllowList = []capi.KeyUsage{
+		capi.UsageKeyEncipherment,
+		capi.UsageDigitalSignature,
+		capi.UsageClientAuth,
+	}
+)
+
+type kubeletReadonlyCSRRequestContent struct {
+	context           context.Context
+	controllerContext *controllerContext
+	csr               *capi.CertificateSigningRequest
+	x509cr            *x509.CertificateRequest
+	autopilotEnabled  bool
+}
+
+type kubeletReadonlyCSRRequestResult struct {
+	result  bool
+	err     error
+	message string
+}
+
+type kubeletReadonlyCSRValidateFunc func(content kubeletReadonlyCSRRequestContent) kubeletReadonlyCSRRequestResult
+
+// readonlyApprover handles approval/denial of CSRs access to read only port
+// There is a list of approvers, requst must get all approval
+type kubeletReadonlyApprover struct {
+	ctx        *controllerContext
+	validators []kubeletReadonlyCsrValidator
+}
+
+func newKubeletReadonlyCSRApprover(ctx *controllerContext) kubeletReadonlyApprover {
+	return kubeletReadonlyApprover{
+		ctx:        ctx,
+		validators: newKubeletReadonlyCsrValidator(),
+	}
+}
+
+func (approver *kubeletReadonlyApprover) handle(ctx context.Context, csr *capi.CertificateSigningRequest) error {
+	if csr == nil || csr.Spec.SignerName != kubeletReadonlyCSRSignerName {
+		return nil
+	}
+	klog.Infof("approver got CSR %q", csr.Name)
+	recordMetric := csrmetrics.ApprovalStartRecorder(kubeletReadonlyCSRRequestMetrics)
+	if len(csr.Status.Certificate) != 0 {
+		return nil
+	}
+	if approved, denied := certificates.GetCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+
+	x509cr, err := certutil.ParseCSR(csr.Spec.Request)
+	if err != nil {
+		recordMetric(csrmetrics.ApprovalStatusParseError)
+		klog.Errorf("unable to parse csr %q: %v", csr.Name, err)
+		return fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+	}
+
+	for _, validator := range approver.validators {
+		recordValidatorMetric := csrmetrics.ApprovalStartRecorder(validator.authFlowLabel)
+		content := kubeletReadonlyCSRRequestContent{
+			context:           ctx,
```

Comment:
create this once?

Reply:
It will differentiate based on authFLowLabel